packages/block-editor/src/components/block-list/hover-area.js

```diff
@@ -1,82 +1,41 @@
 /**
  * WordPress dependencies
  */
-import { Component } from '@wordpress/element';
-import { withSelect } from '@wordpress/data';
+import { useState, useEffect } from '@wordpress/element';
 
-class HoverArea extends Component {
-	constructor() {
-		super( ...arguments );
-		this.state = {
-			hoverArea: null,
-		};
-		this.onMouseLeave = this.onMouseLeave.bind( this );
-		this.onMouseMove = this.onMouseMove.bind( this );
-	}
-
-	componentWillUnmount() {
-		if ( this.props.container ) {
-			this.toggleListeners( this.props.container, false );
-		}
-	}
-
-	componentDidMount() {
-		if ( this.props.container ) {
-			this.toggleListeners( this.props.container );
-		}
-	}
-
-	componentDidUpdate( prevProps ) {
-		if ( prevProps.container === this.props.container ) {
-			return;
-		}
-		if ( prevProps.container ) {
-			this.toggleListeners( prevProps.container, false );
-		}
-		if ( this.props.container ) {
-			this.toggleListeners( this.props.container, true );
-		}
-	}
+const useHoveredArea = ( wrapper ) => {
+	const [ hoveredArea, setHoveredArea ] = useState( null );
 
-	toggleListeners( container, shouldListnerToEvents = true ) {
-		const method = shouldListnerToEvents ? 'addEventListener' : 'removeEventListener';
-		container[ method ]( 'mousemove', this.onMouseMove );
-		container[ method ]( 'mouseleave', this.onMouseLeave );
-	}
-
-	onMouseLeave() {
-		if ( this.state.hoverArea ) {
-			this.setState( { hoverArea: null } );
-		}
-	}
+	useEffect( () => {
+		const onMouseLeave = () => {
+			if ( hoveredArea ) {
+				setHoveredArea( null );
+			}
+		};
 
-	onMouseMove( event ) {
-		const { isRTL, container } = this.props;
-		const { width, left, right } = container.getBoundingClientRect();
+		const onMouseMove = ( event ) => {
+			const { width, left, right } = wrapper.current.getBoundingClientRect();
 
-		let hoverArea = null;
-		if ( ( event.clientX - left ) < width / 3 ) {
-			hoverArea = isRTL ? 'right' : 'left';
-		} else if ( ( right - event.clientX ) < width / 3 ) {
-			hoverArea = isRTL ? 'left' : 'right';
-		}
+			let newHoveredArea = null;
+			if ( ( event.clientX - left ) < width / 3 ) {
+				newHoveredArea = 'left';
+			} else if ( ( right - event.clientX ) < width / 3 ) {
+				newHoveredArea = 'right';
+			}
 
-		if ( hoverArea !== this.state.hoverArea ) {
-			this.setState( { hoverArea } );
-		}
-	}
+			setHoveredArea( newHoveredArea );
+		};
 
-	render() {
-		const { hoverArea } = this.state;
-		const { children } = this.props;
+		wrapper.current.addEventListener( 'mousemove', onMouseMove );
```

Comment:
So, a few questions for my own clarification:

- Even though the wrapper passed from `block.js` is initialized as `null` using `useRef`, we're safe to know that it's a DOM reference, presumably by the fact that `useEffect` callback will occur _after_ the initial render?
- Does `wrapper` need to be a dependency of the `useEffect` or are we okay to assume that it will never change (I assume this will help us to avoid needing to unbind / rebind event handlers too?)

Reply:
> Even though the wrapper passed from block.js is initialized as null using useRef, we're safe to know that it's a DOM reference, presumably by the fact that useEffect callback will occur after the initial render?

yes, also took a look at some react community hooks and a lot of them use a similar technique.

> Does wrapper need to be a dependency of the useEffect or are we okay to assume that it will never change (I assume this will help us to avoid needing to unbind / rebind event handlers too?)

This is the kind of questions I'm still unsure about. My naive thinking would say that leaving it out is better in terms of performance especially because we know that it doesn't change. (Same here, community hooks based on refs don't add it, at least the ones I saw).